// @unique-nft/sdk/uniqueIndexer.cjs
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/uniqueIndexer.ts
var uniqueIndexer_exports = {};
__export(uniqueIndexer_exports, {
  UniqueIndexer: () => UniqueIndexer
});
module.exports = __toCommonJS(uniqueIndexer_exports);

// src/indexer/index.ts
var import_axios = __toESM(require("axios"), 1);
var import_qs = __toESM(require("qs"), 1);

// src/indexer/api.ts
var import_zod = require("zod");
var accounts = {
  zSingleAccountParams: import_zod.z.object({
    address: import_zod.z.string()
  }),
  zAccountRowSelect: import_zod.z.object({
    address: import_zod.z.string(),
    prototype: import_zod.z.string().nullable(),
    mirror: import_zod.z.string(),
    isEthereum: import_zod.z.boolean(),
    isSigner: import_zod.z.boolean().nullable(),
    isSmartContract: import_zod.z.boolean().nullable(),
    createdAtBlockNumber: import_zod.z.number(),
    updatedAtBlockNumber: import_zod.z.number(),
    createdAt: import_zod.z.coerce.date(),
    updatedAt: import_zod.z.coerce.date()
  }),
  zAccountsFilter: import_zod.z.object({
    addressIn: import_zod.z.array(import_zod.z.string()).optional(),
    mirrorIn: import_zod.z.array(import_zod.z.string()).optional(),
    isEthereum: import_zod.z.boolean().optional(),
    isSigner: import_zod.z.boolean().optional(),
    isSmartContract: import_zod.z.boolean().optional(),
    createdTimestampFrom: import_zod.z.coerce.date().optional(),
    createdTimestampTo: import_zod.z.coerce.date().optional(),
    offset: import_zod.z.number().default(0).optional(),
    limit: import_zod.z.number().default(100).optional(),
    orderByAddress: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByCreatedTimestamp: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByCreatedAtBlockNumber: import_zod.z.enum(["asc", "desc"]).optional()
  }),
  zAccountsWithCount: import_zod.z.object({
    items: import_zod.z.array(
      import_zod.z.object({
        address: import_zod.z.string(),
        prototype: import_zod.z.string().nullable(),
        mirror: import_zod.z.string(),
        isEthereum: import_zod.z.boolean(),
        isSigner: import_zod.z.boolean().nullable(),
        isSmartContract: import_zod.z.boolean().nullable(),
        createdAtBlockNumber: import_zod.z.number(),
        updatedAtBlockNumber: import_zod.z.number(),
        createdAt: import_zod.z.coerce.date(),
        updatedAt: import_zod.z.coerce.date()
      })
    ),
    count: import_zod.z.number()
  })
};
var blocks = {
  zSingleBlockParam: import_zod.z.object({
    hashOrNumber: import_zod.z.string()
  }),
  zBlockRowSelect: import_zod.z.object({
    number: import_zod.z.number(),
    hash: import_zod.z.string(),
    parentHash: import_zod.z.string(),
    specVersion: import_zod.z.number(),
    timestamp: import_zod.z.coerce.date(),
    relayParentNumber: import_zod.z.number(),
    relayParentStorageRoot: import_zod.z.string(),
    createdAt: import_zod.z.coerce.date(),
    extrinsics: import_zod.z.array(
      import_zod.z.object({
        hash: import_zod.z.string(),
        blockHash: import_zod.z.string(),
        blockNumber: import_zod.z.number(),
        indexInBlock: import_zod.z.number(),
        signer: import_zod.z.string(),
        section: import_zod.z.string(),
        method: import_zod.z.string(),
        args: import_zod.z.record(import_zod.z.any()),
        asHex: import_zod.z.string(),
        createdAt: import_zod.z.coerce.date(),
        events: import_zod.z.array(
          import_zod.z.object({
            extrinsicHash: import_zod.z.string(),
            blockNumber: import_zod.z.number(),
            section: import_zod.z.string(),
            method: import_zod.z.string(),
            data: import_zod.z.record(import_zod.z.any()),
            asHex: import_zod.z.string(),
            createdAt: import_zod.z.coerce.date()
          })
        ).default([]).optional()
      })
    ).default([]).optional()
  }),
  zBlocksFilter: import_zod.z.object({
    numberFrom: import_zod.z.number().optional(),
    numberTo: import_zod.z.number().optional(),
    relayParentNumberFrom: import_zod.z.number().optional(),
    relayParentNumberTo: import_zod.z.number().optional(),
    timestampFrom: import_zod.z.coerce.date().optional(),
    timestampTo: import_zod.z.coerce.date().optional(),
    limit: import_zod.z.number().default(100).optional(),
    offset: import_zod.z.number().default(0).optional(),
    orderByNumber: import_zod.z.enum(["asc", "desc"]).optional()
  }),
  zBlocksWithCount: import_zod.z.object({
    items: import_zod.z.array(
      import_zod.z.object({
        number: import_zod.z.number(),
        hash: import_zod.z.string(),
        parentHash: import_zod.z.string(),
        specVersion: import_zod.z.number(),
        timestamp: import_zod.z.coerce.date(),
        relayParentNumber: import_zod.z.number(),
        relayParentStorageRoot: import_zod.z.string(),
        createdAt: import_zod.z.coerce.date(),
        extrinsics: import_zod.z.array(
          import_zod.z.object({
            hash: import_zod.z.string(),
            blockHash: import_zod.z.string(),
            blockNumber: import_zod.z.number(),
            indexInBlock: import_zod.z.number(),
            signer: import_zod.z.string(),
            section: import_zod.z.string(),
            method: import_zod.z.string(),
            args: import_zod.z.record(import_zod.z.any()),
            asHex: import_zod.z.string(),
            createdAt: import_zod.z.coerce.date(),
            events: import_zod.z.array(
              import_zod.z.object({
                extrinsicHash: import_zod.z.string(),
                blockNumber: import_zod.z.number(),
                section: import_zod.z.string(),
                method: import_zod.z.string(),
                data: import_zod.z.record(import_zod.z.any()),
                asHex: import_zod.z.string(),
                createdAt: import_zod.z.coerce.date()
              })
            ).default([]).optional()
          })
        ).default([]).optional()
      })
    ),
    count: import_zod.z.number()
  })
};
var coinBalances = {
  zSingleCoinBalanceParam: import_zod.z.object({
    address: import_zod.z.string()
  }),
  zCoinBalanceRowSelect: import_zod.z.object({
    address: import_zod.z.string(),
    available: import_zod.z.string(),
    locked: import_zod.z.string(),
    free: import_zod.z.string(),
    total: import_zod.z.string(),
    reserved: import_zod.z.string(),
    staked: import_zod.z.string(),
    unstaked: import_zod.z.string(),
    canstake: import_zod.z.string(),
    createdAtBlockNumber: import_zod.z.number(),
    updatedAtBlockNumber: import_zod.z.number(),
    fetchedAtBlockNumber: import_zod.z.number(),
    createdAt: import_zod.z.coerce.date(),
    updatedAt: import_zod.z.coerce.date()
  }),
  zCoinBalancesFilter: import_zod.z.object({
    addressIn: import_zod.z.array(import_zod.z.string()).optional(),
    offset: import_zod.z.number().default(0).optional(),
    limit: import_zod.z.number().default(100).optional(),
    availableFrom: import_zod.z.bigint().optional(),
    availableTo: import_zod.z.bigint().optional(),
    lockedFrom: import_zod.z.bigint().optional(),
    lockedTo: import_zod.z.bigint().optional(),
    freeFrom: import_zod.z.bigint().optional(),
    freeTo: import_zod.z.bigint().optional(),
    totalFrom: import_zod.z.bigint().optional(),
    totalTo: import_zod.z.bigint().optional(),
    reservedFrom: import_zod.z.bigint().optional(),
    reservedTo: import_zod.z.bigint().optional(),
    stakedFrom: import_zod.z.bigint().optional(),
    stakedTo: import_zod.z.bigint().optional(),
    unstakedFrom: import_zod.z.bigint().optional(),
    unstakedTo: import_zod.z.bigint().optional(),
    canstakeFrom: import_zod.z.bigint().optional(),
    canstakeTo: import_zod.z.bigint().optional(),
    orderByAddress: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByAvailable: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByLocked: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByFree: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByTotal: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByReserved: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByStaked: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByUnstaked: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByCanstake: import_zod.z.enum(["asc", "desc"]).optional()
  }),
  zCoinBalancesWithCount: import_zod.z.object({
    items: import_zod.z.array(
      import_zod.z.object({
        address: import_zod.z.string(),
        available: import_zod.z.string(),
        locked: import_zod.z.string(),
        free: import_zod.z.string(),
        total: import_zod.z.string(),
        reserved: import_zod.z.string(),
        staked: import_zod.z.string(),
        unstaked: import_zod.z.string(),
        canstake: import_zod.z.string(),
        createdAtBlockNumber: import_zod.z.number(),
        updatedAtBlockNumber: import_zod.z.number(),
        fetchedAtBlockNumber: import_zod.z.number(),
        createdAt: import_zod.z.coerce.date(),
        updatedAt: import_zod.z.coerce.date()
      })
    ),
    count: import_zod.z.number()
  })
};
var coinTransfers = {
  zCoinBalancesTransfersFilter: import_zod.z.object({
    fromIn: import_zod.z.array(import_zod.z.string()).optional(),
    toIn: import_zod.z.array(import_zod.z.string()).optional(),
    amountFrom: import_zod.z.bigint().optional(),
    amountTo: import_zod.z.bigint().optional(),
    orderByFrom: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByTo: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByAmount: import_zod.z.enum(["asc", "desc"]).optional(),
    offset: import_zod.z.number().default(0).optional(),
    limit: import_zod.z.number().default(100).optional()
  }),
  zCoinBalancesTransfersWithCount: import_zod.z.object({
    items: import_zod.z.array(
      import_zod.z.object({
        id: import_zod.z.number(),
        from: import_zod.z.string(),
        to: import_zod.z.string(),
        amount: import_zod.z.string(),
        eventId: import_zod.z.number(),
        event: import_zod.z.object({
          blockNumber: import_zod.z.number(),
          extrinsicHash: import_zod.z.string(),
          extrinsic: import_zod.z.object({
            indexInBlock: import_zod.z.number(),
            section: import_zod.z.string().optional(),
            method: import_zod.z.string().optional(),
            signer: import_zod.z.string().optional()
          })
        }).nullable()
      })
    ),
    count: import_zod.z.number()
  })
};
var collections = {
  zCollectionsFilter: import_zod.z.object({
    mode: import_zod.z.enum(["NFT", "Fungible", "RFT"]).optional(),
    collectionIdIn: import_zod.z.array(import_zod.z.string()).optional(),
    nameLike: import_zod.z.string().optional(),
    descriptionLike: import_zod.z.string().optional(),
    ownerIn: import_zod.z.array(import_zod.z.string()).optional(),
    adminIn: import_zod.z.array(import_zod.z.string()).optional(),
    sponsorIn: import_zod.z.array(import_zod.z.string()).optional(),
    isBurned: import_zod.z.boolean().default(false).optional(),
    createdTimestampFrom: import_zod.z.coerce.date().optional(),
    createdTimestampTo: import_zod.z.coerce.date().optional(),
    updatedTimestampFrom: import_zod.z.coerce.date().optional(),
    updatedTimestampTo: import_zod.z.coerce.date().optional(),
    offset: import_zod.z.number().default(0).optional(),
    limit: import_zod.z.number().default(10).optional(),
    orderByCollectionId: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByName: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByDescription: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByOwner: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByUpdatedAtBlockNumber: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByCreatedAtBlockNumber: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByLastTokenId: import_zod.z.enum(["asc", "desc"]).optional()
  }),
  zCollectionsWithCount: import_zod.z.object({
    items: import_zod.z.array(
      import_zod.z.object({
        collectionId: import_zod.z.number(),
        collectionAddress: import_zod.z.string(),
        owner: import_zod.z.string(),
        adminList: import_zod.z.array(import_zod.z.string()),
        mode: import_zod.z.string(),
        fungiblePieces: import_zod.z.number().nullable(),
        name: import_zod.z.string(),
        description: import_zod.z.string(),
        tokenPrefix: import_zod.z.string(),
        sponsor: import_zod.z.string().nullable(),
        isSponsorshipConfirmed: import_zod.z.boolean().nullable(),
        lastTokenId: import_zod.z.number(),
        limits: import_zod.z.object({
          account_token_ownership_limit: import_zod.z.number().nullable(),
          owner_can_destroy: import_zod.z.boolean().nullable(),
          owner_can_transfer: import_zod.z.boolean().nullable(),
          sponsor_approve_timeout: import_zod.z.number().nullable(),
          sponsor_transfer_timeout: import_zod.z.number().nullable(),
          sponsored_data_rate_limit: import_zod.z.union([
            import_zod.z.literal("SponsoringDisabled"),
            import_zod.z.object({
              Blocks: import_zod.z.number()
            })
          ]).nullable(),
          sponsored_data_size: import_zod.z.number().nullable(),
          token_limit: import_zod.z.number().nullable(),
          transfers_enabled: import_zod.z.boolean().nullable()
        }),
        decodedLimits: import_zod.z.record(
          import_zod.z.object({
            key: import_zod.z.string(),
            value: import_zod.z.union([
              import_zod.z.number(),
              import_zod.z.boolean(),
              import_zod.z.string(),
              import_zod.z.object({
                Blocks: import_zod.z.number()
              })
            ]),
            isDefaultValue: import_zod.z.boolean()
          })
        ),
        permissions: import_zod.z.object({
          access: import_zod.z.union([import_zod.z.literal("Normal"), import_zod.z.literal("AllowList")]),
          mint_mode: import_zod.z.boolean(),
          nesting: import_zod.z.object({
            token_owner: import_zod.z.boolean(),
            collection_admin: import_zod.z.boolean(),
            restricted: import_zod.z.union([import_zod.z.null(), import_zod.z.array(import_zod.z.number())])
          })
        }),
        tokenPropertyPermissions: import_zod.z.array(
          import_zod.z.object({
            key: import_zod.z.string(),
            keyHex: import_zod.z.string(),
            permission: import_zod.z.object({
              mutable: import_zod.z.boolean(),
              collection_admin: import_zod.z.boolean(),
              token_owner: import_zod.z.boolean()
            })
          })
        ),
        properties: import_zod.z.array(
          import_zod.z.object({
            key: import_zod.z.string(),
            keyHex: import_zod.z.string(),
            value: import_zod.z.string(),
            valueHex: import_zod.z.string().optional()
          })
        ),
        propertiesMap: import_zod.z.record(
          import_zod.z.object({
            key: import_zod.z.string(),
            keyHex: import_zod.z.string(),
            value: import_zod.z.string(),
            valueHex: import_zod.z.string().optional()
          })
        ),
        readOnly: import_zod.z.boolean(),
        foreign: import_zod.z.boolean().nullable(),
        erc721metadata: import_zod.z.boolean().nullable(),
        schemaName: import_zod.z.string().nullable(),
        schemaVersion: import_zod.z.string().nullable(),
        originalSchemaVersion: import_zod.z.string().nullable(),
        coverImage: import_zod.z.object({
          url: import_zod.z.string(),
          details: import_zod.z.object({
            name: import_zod.z.string().optional(),
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
            bytes: import_zod.z.number().optional(),
            format: import_zod.z.string().optional(),
            sha256: import_zod.z.string().optional(),
            width: import_zod.z.number().optional(),
            height: import_zod.z.number().optional(),
            order: import_zod.z.number().optional()
          }).optional(),
          thumbnail: import_zod.z.object({
            url: import_zod.z.string(),
            details: import_zod.z.object({
              name: import_zod.z.string().optional(),
              type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
              bytes: import_zod.z.number().optional(),
              format: import_zod.z.string().optional(),
              sha256: import_zod.z.string().optional(),
              width: import_zod.z.number().optional(),
              height: import_zod.z.number().optional(),
              order: import_zod.z.number().optional()
            }).optional()
          }).optional()
        }).optional().nullable(),
        defaultTokenImage: import_zod.z.object({
          url: import_zod.z.string(),
          details: import_zod.z.object({
            name: import_zod.z.string().optional(),
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
            bytes: import_zod.z.number().optional(),
            format: import_zod.z.string().optional(),
            sha256: import_zod.z.string().optional(),
            width: import_zod.z.number().optional(),
            height: import_zod.z.number().optional(),
            order: import_zod.z.number().optional()
          }).optional(),
          thumbnail: import_zod.z.object({
            url: import_zod.z.string(),
            details: import_zod.z.object({
              name: import_zod.z.string().optional(),
              type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
              bytes: import_zod.z.number().optional(),
              format: import_zod.z.string().optional(),
              sha256: import_zod.z.string().optional(),
              width: import_zod.z.number().optional(),
              height: import_zod.z.number().optional(),
              order: import_zod.z.number().optional()
            }).optional()
          }).optional()
        }).optional().nullable(),
        potentialAttributes: import_zod.z.array(
          import_zod.z.object({
            trait_type: import_zod.z.string(),
            display_type: import_zod.z.string().optional(),
            values: import_zod.z.array(import_zod.z.union([import_zod.z.string(), import_zod.z.number()])).optional()
          })
        ).optional().nullable(),
        customizingSlots: import_zod.z.record(
          import_zod.z.object({
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]),
            collections: import_zod.z.array(import_zod.z.union([import_zod.z.number(), import_zod.z.string()])).optional(),
            name: import_zod.z.string().optional(),
            image_overlay_specs: import_zod.z.object({
              layer: import_zod.z.number().optional(),
              order_in_layer: import_zod.z.number().optional(),
              offset: import_zod.z.object({
                x: import_zod.z.number().optional(),
                y: import_zod.z.number().optional()
              }).optional(),
              opacity: import_zod.z.number().optional(),
              rotation: import_zod.z.number().optional(),
              scale: import_zod.z.object({
                x: import_zod.z.number().optional(),
                y: import_zod.z.number().optional(),
                unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
              }).optional(),
              anchor_point: import_zod.z.object({
                x: import_zod.z.number(),
                y: import_zod.z.number()
              }).optional(),
              parent_anchor_point: import_zod.z.object({
                x: import_zod.z.number(),
                y: import_zod.z.number()
              }).optional()
            }).optional()
          })
        ).optional().nullable(),
        customizes: import_zod.z.array(import_zod.z.union([import_zod.z.number(), import_zod.z.string()])).optional().nullable(),
        royalties: import_zod.z.array(
          import_zod.z.object({
            address: import_zod.z.string(),
            percent: import_zod.z.number(),
            isPrimaryOnly: import_zod.z.boolean().optional()
          })
        ).optional().nullable(),
        collectionInfoError: import_zod.z.string().nullable(),
        isBurned: import_zod.z.boolean(),
        fetchedAtBlockNumber: import_zod.z.number(),
        updatedAtBlockNumber: import_zod.z.number(),
        updatedAt: import_zod.z.coerce.date(),
        createdAt: import_zod.z.coerce.date(),
        createdAtBlockNumber: import_zod.z.number()
      })
    ),
    count: import_zod.z.number()
  }),
  zParsedCollectionRowSelect: import_zod.z.object({
    collectionId: import_zod.z.number(),
    collectionAddress: import_zod.z.string(),
    owner: import_zod.z.string(),
    adminList: import_zod.z.array(import_zod.z.string()),
    mode: import_zod.z.string(),
    fungiblePieces: import_zod.z.number().nullable(),
    name: import_zod.z.string(),
    description: import_zod.z.string(),
    tokenPrefix: import_zod.z.string(),
    sponsor: import_zod.z.string().nullable(),
    isSponsorshipConfirmed: import_zod.z.boolean().nullable(),
    lastTokenId: import_zod.z.number(),
    limits: import_zod.z.object({
      account_token_ownership_limit: import_zod.z.number().nullable(),
      owner_can_destroy: import_zod.z.boolean().nullable(),
      owner_can_transfer: import_zod.z.boolean().nullable(),
      sponsor_approve_timeout: import_zod.z.number().nullable(),
      sponsor_transfer_timeout: import_zod.z.number().nullable(),
      sponsored_data_rate_limit: import_zod.z.union([
        import_zod.z.literal("SponsoringDisabled"),
        import_zod.z.object({
          Blocks: import_zod.z.number()
        })
      ]).nullable(),
      sponsored_data_size: import_zod.z.number().nullable(),
      token_limit: import_zod.z.number().nullable(),
      transfers_enabled: import_zod.z.boolean().nullable()
    }),
    decodedLimits: import_zod.z.record(
      import_zod.z.object({
        key: import_zod.z.string(),
        value: import_zod.z.union([
          import_zod.z.number(),
          import_zod.z.boolean(),
          import_zod.z.string(),
          import_zod.z.object({
            Blocks: import_zod.z.number()
          })
        ]),
        isDefaultValue: import_zod.z.boolean()
      })
    ),
    permissions: import_zod.z.object({
      access: import_zod.z.union([import_zod.z.literal("Normal"), import_zod.z.literal("AllowList")]),
      mint_mode: import_zod.z.boolean(),
      nesting: import_zod.z.object({
        token_owner: import_zod.z.boolean(),
        collection_admin: import_zod.z.boolean(),
        restricted: import_zod.z.union([import_zod.z.null(), import_zod.z.array(import_zod.z.number())])
      })
    }),
    tokenPropertyPermissions: import_zod.z.array(
      import_zod.z.object({
        key: import_zod.z.string(),
        keyHex: import_zod.z.string(),
        permission: import_zod.z.object({
          mutable: import_zod.z.boolean(),
          collection_admin: import_zod.z.boolean(),
          token_owner: import_zod.z.boolean()
        })
      })
    ),
    properties: import_zod.z.array(
      import_zod.z.object({
        key: import_zod.z.string(),
        keyHex: import_zod.z.string(),
        value: import_zod.z.string(),
        valueHex: import_zod.z.string().optional()
      })
    ),
    propertiesMap: import_zod.z.record(
      import_zod.z.object({
        key: import_zod.z.string(),
        keyHex: import_zod.z.string(),
        value: import_zod.z.string(),
        valueHex: import_zod.z.string().optional()
      })
    ),
    readOnly: import_zod.z.boolean(),
    foreign: import_zod.z.boolean().nullable(),
    erc721metadata: import_zod.z.boolean().nullable(),
    schemaName: import_zod.z.string().nullable(),
    schemaVersion: import_zod.z.string().nullable(),
    originalSchemaVersion: import_zod.z.string().nullable(),
    coverImage: import_zod.z.object({
      url: import_zod.z.string(),
      details: import_zod.z.object({
        name: import_zod.z.string().optional(),
        type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
        bytes: import_zod.z.number().optional(),
        format: import_zod.z.string().optional(),
        sha256: import_zod.z.string().optional(),
        width: import_zod.z.number().optional(),
        height: import_zod.z.number().optional(),
        order: import_zod.z.number().optional()
      }).optional(),
      thumbnail: import_zod.z.object({
        url: import_zod.z.string(),
        details: import_zod.z.object({
          name: import_zod.z.string().optional(),
          type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
          bytes: import_zod.z.number().optional(),
          format: import_zod.z.string().optional(),
          sha256: import_zod.z.string().optional(),
          width: import_zod.z.number().optional(),
          height: import_zod.z.number().optional(),
          order: import_zod.z.number().optional()
        }).optional()
      }).optional()
    }).optional().nullable(),
    defaultTokenImage: import_zod.z.object({
      url: import_zod.z.string(),
      details: import_zod.z.object({
        name: import_zod.z.string().optional(),
        type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
        bytes: import_zod.z.number().optional(),
        format: import_zod.z.string().optional(),
        sha256: import_zod.z.string().optional(),
        width: import_zod.z.number().optional(),
        height: import_zod.z.number().optional(),
        order: import_zod.z.number().optional()
      }).optional(),
      thumbnail: import_zod.z.object({
        url: import_zod.z.string(),
        details: import_zod.z.object({
          name: import_zod.z.string().optional(),
          type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
          bytes: import_zod.z.number().optional(),
          format: import_zod.z.string().optional(),
          sha256: import_zod.z.string().optional(),
          width: import_zod.z.number().optional(),
          height: import_zod.z.number().optional(),
          order: import_zod.z.number().optional()
        }).optional()
      }).optional()
    }).optional().nullable(),
    potentialAttributes: import_zod.z.array(
      import_zod.z.object({
        trait_type: import_zod.z.string(),
        display_type: import_zod.z.string().optional(),
        values: import_zod.z.array(import_zod.z.union([import_zod.z.string(), import_zod.z.number()])).optional()
      })
    ).optional().nullable(),
    customizingSlots: import_zod.z.record(
      import_zod.z.object({
        type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]),
        collections: import_zod.z.array(import_zod.z.union([import_zod.z.number(), import_zod.z.string()])).optional(),
        name: import_zod.z.string().optional(),
        image_overlay_specs: import_zod.z.object({
          layer: import_zod.z.number().optional(),
          order_in_layer: import_zod.z.number().optional(),
          offset: import_zod.z.object({
            x: import_zod.z.number().optional(),
            y: import_zod.z.number().optional()
          }).optional(),
          opacity: import_zod.z.number().optional(),
          rotation: import_zod.z.number().optional(),
          scale: import_zod.z.object({
            x: import_zod.z.number().optional(),
            y: import_zod.z.number().optional(),
            unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
          }).optional(),
          anchor_point: import_zod.z.object({
            x: import_zod.z.number(),
            y: import_zod.z.number()
          }).optional(),
          parent_anchor_point: import_zod.z.object({
            x: import_zod.z.number(),
            y: import_zod.z.number()
          }).optional()
        }).optional()
      })
    ).optional().nullable(),
    customizes: import_zod.z.array(import_zod.z.union([import_zod.z.number(), import_zod.z.string()])).optional().nullable(),
    royalties: import_zod.z.array(
      import_zod.z.object({
        address: import_zod.z.string(),
        percent: import_zod.z.number(),
        isPrimaryOnly: import_zod.z.boolean().optional()
      })
    ).optional().nullable(),
    collectionInfoError: import_zod.z.string().nullable(),
    isBurned: import_zod.z.boolean(),
    fetchedAtBlockNumber: import_zod.z.number(),
    updatedAtBlockNumber: import_zod.z.number(),
    updatedAt: import_zod.z.coerce.date(),
    createdAt: import_zod.z.coerce.date(),
    createdAtBlockNumber: import_zod.z.number()
  }),
  zSingleCollectionParam: import_zod.z.object({
    collectionId: import_zod.z.string()
  })
};
var events = {
  zEventsFilter: import_zod.z.object({
    sectionIn: import_zod.z.array(import_zod.z.string()).optional(),
    methodIn: import_zod.z.array(import_zod.z.string()).optional(),
    address: import_zod.z.string().optional(),
    blockNumberFrom: import_zod.z.number().optional(),
    blockNumberTo: import_zod.z.number().optional(),
    limit: import_zod.z.number().default(100).optional(),
    offset: import_zod.z.number().default(0).optional(),
    orderByBlockNumber: import_zod.z.enum(["asc", "desc"]).optional()
  }),
  zEventWithCount: import_zod.z.object({
    items: import_zod.z.array(
      import_zod.z.object({
        extrinsicHash: import_zod.z.string(),
        blockNumber: import_zod.z.number(),
        blockTimestamp: import_zod.z.number(),
        section: import_zod.z.string(),
        method: import_zod.z.string(),
        data: import_zod.z.record(import_zod.z.any()),
        asHex: import_zod.z.string(),
        createdAt: import_zod.z.coerce.date()
      })
    ),
    count: import_zod.z.number()
  }),
  zSectionMethodArray: import_zod.z.array(
    import_zod.z.object({
      section: import_zod.z.string(),
      method: import_zod.z.string()
    })
  )
};
var extrinsics = {
  zSingleExtrinsicParam: import_zod.z.object({
    hash: import_zod.z.string()
  }),
  zExtrinsicsFilter: import_zod.z.object({
    sectionIn: import_zod.z.array(import_zod.z.string()).optional(),
    methodIn: import_zod.z.array(import_zod.z.string()).optional(),
    signerIn: import_zod.z.array(import_zod.z.string()).optional(),
    blockNumberFrom: import_zod.z.number().optional(),
    blockNumberTo: import_zod.z.number().optional(),
    limit: import_zod.z.number().default(100).optional(),
    offset: import_zod.z.number().default(0).optional(),
    orderByBlockNumber: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByIndex: import_zod.z.enum(["asc", "desc"]).optional()
  }),
  zExtrinsicRowSelect: import_zod.z.object({
    hash: import_zod.z.string(),
    blockHash: import_zod.z.string(),
    blockNumber: import_zod.z.number(),
    blockTimestamp: import_zod.z.number(),
    indexInBlock: import_zod.z.number(),
    signer: import_zod.z.string(),
    section: import_zod.z.string(),
    method: import_zod.z.string(),
    args: import_zod.z.record(import_zod.z.any()),
    asHex: import_zod.z.string(),
    createdAt: import_zod.z.coerce.date(),
    events: import_zod.z.array(
      import_zod.z.object({
        section: import_zod.z.string(),
        method: import_zod.z.string(),
        data: import_zod.z.record(import_zod.z.any())
      })
    ).default([]).optional()
  }),
  zExtrinsicsWithCount: import_zod.z.object({
    items: import_zod.z.array(
      import_zod.z.object({
        hash: import_zod.z.string(),
        blockHash: import_zod.z.string(),
        blockNumber: import_zod.z.number(),
        blockTimestamp: import_zod.z.number(),
        indexInBlock: import_zod.z.number(),
        signer: import_zod.z.string(),
        section: import_zod.z.string(),
        method: import_zod.z.string(),
        args: import_zod.z.record(import_zod.z.any()),
        asHex: import_zod.z.string(),
        createdAt: import_zod.z.coerce.date(),
        events: import_zod.z.array(
          import_zod.z.object({
            section: import_zod.z.string(),
            method: import_zod.z.string(),
            data: import_zod.z.record(import_zod.z.any())
          })
        ).default([]).optional()
      })
    ),
    count: import_zod.z.number()
  })
};
var health = {
  zHealth: import_zod.z.object({
    status: import_zod.z.enum(["ok", "error"]),
    message: import_zod.z.string().optional()
  })
};
var parsedNfts = {
  zParsedNftRowSelect: import_zod.z.object({
    key: import_zod.z.string(),
    collectionId: import_zod.z.number(),
    tokenId: import_zod.z.number(),
    collectionAddress: import_zod.z.string(),
    tokenAddress: import_zod.z.string(),
    owner: import_zod.z.string(),
    topmostOwner: import_zod.z.string().nullable(),
    isBundle: import_zod.z.boolean(),
    properties: import_zod.z.array(
      import_zod.z.object({
        key: import_zod.z.string(),
        keyHex: import_zod.z.string(),
        value: import_zod.z.string(),
        valueHex: import_zod.z.string().optional()
      })
    ).nullable(),
    propertiesMap: import_zod.z.record(
      import_zod.z.object({
        key: import_zod.z.string(),
        keyHex: import_zod.z.string(),
        value: import_zod.z.string(),
        valueHex: import_zod.z.string().optional()
      })
    ).nullable(),
    schemaName: import_zod.z.string().nullable(),
    schemaVersion: import_zod.z.string().nullable(),
    originalSchemaVersion: import_zod.z.string().nullable(),
    name: import_zod.z.string().nullable(),
    description: import_zod.z.string().nullable(),
    image: import_zod.z.string().nullable(),
    imageDetails: import_zod.z.object({
      name: import_zod.z.string().optional(),
      type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
      bytes: import_zod.z.number().optional(),
      format: import_zod.z.string().optional(),
      sha256: import_zod.z.string().optional(),
      width: import_zod.z.number().optional(),
      height: import_zod.z.number().optional(),
      order: import_zod.z.number().optional()
    }).optional().nullable(),
    attributes: import_zod.z.array(
      import_zod.z.object({
        trait_type: import_zod.z.string(),
        value: import_zod.z.union([import_zod.z.string(), import_zod.z.number()]),
        display_type: import_zod.z.string().optional()
      })
    ).optional().nullable(),
    media: import_zod.z.record(
      import_zod.z.object({
        type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]),
        url: import_zod.z.string(),
        name: import_zod.z.string().optional(),
        details: import_zod.z.object({
          name: import_zod.z.string().optional(),
          type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
          bytes: import_zod.z.number().optional(),
          format: import_zod.z.string().optional(),
          sha256: import_zod.z.string().optional(),
          width: import_zod.z.number().optional(),
          height: import_zod.z.number().optional(),
          order: import_zod.z.number().optional(),
          duration: import_zod.z.number().optional(),
          codecs: import_zod.z.array(import_zod.z.string()).optional(),
          loop: import_zod.z.boolean().optional()
        }).optional(),
        thumbnail: import_zod.z.object({
          url: import_zod.z.string(),
          details: import_zod.z.object({
            name: import_zod.z.string().optional(),
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
            bytes: import_zod.z.number().optional(),
            format: import_zod.z.string().optional(),
            sha256: import_zod.z.string().optional(),
            width: import_zod.z.number().optional(),
            height: import_zod.z.number().optional(),
            order: import_zod.z.number().optional()
          }).optional()
        }).optional(),
        poster: import_zod.z.object({
          url: import_zod.z.string(),
          details: import_zod.z.object({
            name: import_zod.z.string().optional(),
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
            bytes: import_zod.z.number().optional(),
            format: import_zod.z.string().optional(),
            sha256: import_zod.z.string().optional(),
            width: import_zod.z.number().optional(),
            height: import_zod.z.number().optional(),
            order: import_zod.z.number().optional()
          }).optional()
        }).optional()
      })
    ).optional().nullable(),
    royalties: import_zod.z.array(
      import_zod.z.object({
        address: import_zod.z.string(),
        percent: import_zod.z.number(),
        isPrimaryOnly: import_zod.z.boolean().optional()
      })
    ).optional().nullable(),
    customizing: import_zod.z.object({
      self: import_zod.z.object({
        type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]),
        url: import_zod.z.string(),
        name: import_zod.z.string().optional(),
        details: import_zod.z.object({
          name: import_zod.z.string().optional(),
          type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
          bytes: import_zod.z.number().optional(),
          format: import_zod.z.string().optional(),
          sha256: import_zod.z.string().optional(),
          width: import_zod.z.number().optional(),
          height: import_zod.z.number().optional(),
          order: import_zod.z.number().optional(),
          duration: import_zod.z.number().optional(),
          codecs: import_zod.z.array(import_zod.z.string()).optional(),
          loop: import_zod.z.boolean().optional()
        }).optional(),
        image_overlay_specs: import_zod.z.object({
          layer: import_zod.z.number().optional(),
          order_in_layer: import_zod.z.number().optional(),
          offset: import_zod.z.object({
            x: import_zod.z.number().optional(),
            y: import_zod.z.number().optional()
          }).optional(),
          opacity: import_zod.z.number().optional(),
          rotation: import_zod.z.number().optional(),
          scale: import_zod.z.object({
            x: import_zod.z.number().optional(),
            y: import_zod.z.number().optional(),
            unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
          }).optional(),
          anchor_point: import_zod.z.object({
            x: import_zod.z.number(),
            y: import_zod.z.number()
          }).optional(),
          parent_anchor_point: import_zod.z.object({
            x: import_zod.z.number(),
            y: import_zod.z.number()
          }).optional()
        }).optional(),
        placeholder: import_zod.z.object({
          url: import_zod.z.string(),
          details: import_zod.z.object({
            name: import_zod.z.string().optional(),
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
            bytes: import_zod.z.number().optional(),
            format: import_zod.z.string().optional(),
            sha256: import_zod.z.string().optional(),
            width: import_zod.z.number().optional(),
            height: import_zod.z.number().optional(),
            order: import_zod.z.number().optional()
          }).optional()
        }).optional(),
        tag: import_zod.z.string()
      }),
      slots: import_zod.z.record(
        import_zod.z.object({
          type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]),
          collections: import_zod.z.array(import_zod.z.union([import_zod.z.number(), import_zod.z.string()])).optional(),
          name: import_zod.z.string().optional(),
          image_overlay_specs: import_zod.z.object({
            layer: import_zod.z.number().optional(),
            order_in_layer: import_zod.z.number().optional(),
            offset: import_zod.z.object({
              x: import_zod.z.number().optional(),
              y: import_zod.z.number().optional()
            }).optional(),
            opacity: import_zod.z.number().optional(),
            rotation: import_zod.z.number().optional(),
            scale: import_zod.z.object({
              x: import_zod.z.number().optional(),
              y: import_zod.z.number().optional(),
              unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
            }).optional(),
            anchor_point: import_zod.z.object({
              x: import_zod.z.number(),
              y: import_zod.z.number()
            }).optional(),
            parent_anchor_point: import_zod.z.object({
              x: import_zod.z.number(),
              y: import_zod.z.number()
            }).optional()
          }).optional()
        })
      ).optional(),
      mutator_reactions: import_zod.z.record(
        import_zod.z.object({
          layer: import_zod.z.number().optional().optional(),
          order_in_layer: import_zod.z.number().optional().optional(),
          offset: import_zod.z.object({
            x: import_zod.z.number().optional(),
            y: import_zod.z.number().optional()
          }).optional().optional(),
          opacity: import_zod.z.number().optional().optional(),
          rotation: import_zod.z.number().optional().optional(),
          scale: import_zod.z.object({
            x: import_zod.z.number().optional(),
            y: import_zod.z.number().optional(),
            unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
          }).optional().optional(),
          anchor_point: import_zod.z.object({
            x: import_zod.z.number(),
            y: import_zod.z.number()
          }).optional().optional(),
          parent_anchor_point: import_zod.z.object({
            x: import_zod.z.number(),
            y: import_zod.z.number()
          }).optional().optional(),
          url: import_zod.z.string().optional(),
          details: import_zod.z.object({
            name: import_zod.z.string().optional(),
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
            bytes: import_zod.z.number().optional(),
            format: import_zod.z.string().optional(),
            sha256: import_zod.z.string().optional(),
            width: import_zod.z.number().optional(),
            height: import_zod.z.number().optional(),
            order: import_zod.z.number().optional()
          }).optional()
        })
      ).optional(),
      mutators: import_zod.z.array(import_zod.z.string()).optional()
    }).optional().nullable(),
    customizingOverrides: import_zod.z.object({
      self: import_zod.z.object({
        type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
        url: import_zod.z.string().optional(),
        name: import_zod.z.string().optional().optional(),
        details: import_zod.z.object({
          name: import_zod.z.string().optional(),
          type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
          bytes: import_zod.z.number().optional(),
          format: import_zod.z.string().optional(),
          sha256: import_zod.z.string().optional(),
          width: import_zod.z.number().optional(),
          height: import_zod.z.number().optional(),
          order: import_zod.z.number().optional(),
          duration: import_zod.z.number().optional(),
          codecs: import_zod.z.array(import_zod.z.string()).optional(),
          loop: import_zod.z.boolean().optional()
        }).optional().optional(),
        image_overlay_specs: import_zod.z.object({
          layer: import_zod.z.number().optional(),
          order_in_layer: import_zod.z.number().optional(),
          offset: import_zod.z.object({
            x: import_zod.z.number().optional(),
            y: import_zod.z.number().optional()
          }).optional(),
          opacity: import_zod.z.number().optional(),
          rotation: import_zod.z.number().optional(),
          scale: import_zod.z.object({
            x: import_zod.z.number().optional(),
            y: import_zod.z.number().optional(),
            unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
          }).optional(),
          anchor_point: import_zod.z.object({
            x: import_zod.z.number(),
            y: import_zod.z.number()
          }).optional(),
          parent_anchor_point: import_zod.z.object({
            x: import_zod.z.number(),
            y: import_zod.z.number()
          }).optional()
        }).optional().optional(),
        placeholder: import_zod.z.object({
          url: import_zod.z.string(),
          details: import_zod.z.object({
            name: import_zod.z.string().optional(),
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
            bytes: import_zod.z.number().optional(),
            format: import_zod.z.string().optional(),
            sha256: import_zod.z.string().optional(),
            width: import_zod.z.number().optional(),
            height: import_zod.z.number().optional(),
            order: import_zod.z.number().optional()
          }).optional()
        }).optional().optional(),
        tag: import_zod.z.string().optional()
      }).optional(),
      slots: import_zod.z.record(
        import_zod.z.object({
          type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
          collections: import_zod.z.array(import_zod.z.union([import_zod.z.number(), import_zod.z.string()])).optional().optional(),
          name: import_zod.z.string().optional().optional(),
          image_overlay_specs: import_zod.z.object({
            layer: import_zod.z.number().optional(),
            order_in_layer: import_zod.z.number().optional(),
            offset: import_zod.z.object({
              x: import_zod.z.number().optional(),
              y: import_zod.z.number().optional()
            }).optional(),
            opacity: import_zod.z.number().optional(),
            rotation: import_zod.z.number().optional(),
            scale: import_zod.z.object({
              x: import_zod.z.number().optional(),
              y: import_zod.z.number().optional(),
              unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
            }).optional(),
            anchor_point: import_zod.z.object({
              x: import_zod.z.number(),
              y: import_zod.z.number()
            }).optional(),
            parent_anchor_point: import_zod.z.object({
              x: import_zod.z.number(),
              y: import_zod.z.number()
            }).optional()
          }).optional().optional()
        })
      ).optional(),
      mutator_reactions: import_zod.z.record(
        import_zod.z.object({
          layer: import_zod.z.number().optional().optional().optional(),
          order_in_layer: import_zod.z.number().optional().optional().optional(),
          offset: import_zod.z.object({
            x: import_zod.z.number().optional(),
            y: import_zod.z.number().optional()
          }).optional().optional().optional(),
          opacity: import_zod.z.number().optional().optional().optional(),
          rotation: import_zod.z.number().optional().optional().optional(),
          scale: import_zod.z.object({
            x: import_zod.z.number().optional(),
            y: import_zod.z.number().optional(),
            unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
          }).optional().optional().optional(),
          anchor_point: import_zod.z.object({
            x: import_zod.z.number(),
            y: import_zod.z.number()
          }).optional().optional().optional(),
          parent_anchor_point: import_zod.z.object({
            x: import_zod.z.number(),
            y: import_zod.z.number()
          }).optional().optional().optional(),
          url: import_zod.z.string().optional().optional(),
          details: import_zod.z.object({
            name: import_zod.z.string().optional(),
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
            bytes: import_zod.z.number().optional(),
            format: import_zod.z.string().optional(),
            sha256: import_zod.z.string().optional(),
            width: import_zod.z.number().optional(),
            height: import_zod.z.number().optional(),
            order: import_zod.z.number().optional()
          }).optional().optional()
        })
      ).optional(),
      mutators: import_zod.z.array(import_zod.z.string()).optional()
    }).optional().nullable(),
    animationUrl: import_zod.z.string().nullable(),
    animationDetails: import_zod.z.object({
      name: import_zod.z.string().optional(),
      type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
      bytes: import_zod.z.number().optional(),
      format: import_zod.z.string().optional(),
      sha256: import_zod.z.string().optional(),
      width: import_zod.z.number().optional(),
      height: import_zod.z.number().optional(),
      order: import_zod.z.number().optional()
    }).optional().nullable(),
    youtubeUrl: import_zod.z.string().nullable(),
    createdBy: import_zod.z.string().nullable(),
    backgroundColor: import_zod.z.string().nullable(),
    externalUrl: import_zod.z.string().nullable(),
    locale: import_zod.z.string().nullable(),
    tokenInfoError: import_zod.z.string().nullable(),
    isBurned: import_zod.z.boolean(),
    fetchedAtBlockNumber: import_zod.z.number(),
    updatedAtBlockNumber: import_zod.z.number(),
    updatedAt: import_zod.z.coerce.date(),
    createdAtBlockNumber: import_zod.z.number(),
    createdAt: import_zod.z.coerce.date()
  }),
  zParsedNftsFilter: import_zod.z.object({
    collectionIdIn: import_zod.z.array(import_zod.z.string()).optional(),
    tokenIdIn: import_zod.z.array(import_zod.z.number()).optional(),
    ownerIn: import_zod.z.array(import_zod.z.string()).optional(),
    topmostOwnerIn: import_zod.z.array(import_zod.z.string()).optional(),
    royaltyRecipientIn: import_zod.z.array(import_zod.z.string()).optional(),
    isBurned: import_zod.z.boolean().optional(),
    isBundle: import_zod.z.boolean().optional(),
    attributeValueIn: import_zod.z.array(import_zod.z.string()).optional(),
    attributeTraitTypeIn: import_zod.z.array(import_zod.z.string()).optional(),
    createdTimestampFrom: import_zod.z.coerce.date().optional(),
    createdTimestampTo: import_zod.z.coerce.date().optional(),
    updatedTimestampFrom: import_zod.z.coerce.date().optional(),
    updatedTimestampTo: import_zod.z.coerce.date().optional(),
    offset: import_zod.z.number().default(0).optional(),
    limit: import_zod.z.number().default(10).optional(),
    orderByCollectionId: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByTokenId: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByCreatedAtBlockNumber: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByOwner: import_zod.z.enum(["asc", "desc"]).optional()
  }),
  zParsedNftsWithCount: import_zod.z.object({
    items: import_zod.z.array(
      import_zod.z.object({
        key: import_zod.z.string(),
        collectionId: import_zod.z.number(),
        tokenId: import_zod.z.number(),
        collectionAddress: import_zod.z.string(),
        tokenAddress: import_zod.z.string(),
        owner: import_zod.z.string(),
        topmostOwner: import_zod.z.string().nullable(),
        isBundle: import_zod.z.boolean(),
        properties: import_zod.z.array(
          import_zod.z.object({
            key: import_zod.z.string(),
            keyHex: import_zod.z.string(),
            value: import_zod.z.string(),
            valueHex: import_zod.z.string().optional()
          })
        ).nullable(),
        propertiesMap: import_zod.z.record(
          import_zod.z.object({
            key: import_zod.z.string(),
            keyHex: import_zod.z.string(),
            value: import_zod.z.string(),
            valueHex: import_zod.z.string().optional()
          })
        ).nullable(),
        schemaName: import_zod.z.string().nullable(),
        schemaVersion: import_zod.z.string().nullable(),
        originalSchemaVersion: import_zod.z.string().nullable(),
        name: import_zod.z.string().nullable(),
        description: import_zod.z.string().nullable(),
        image: import_zod.z.string().nullable(),
        imageDetails: import_zod.z.object({
          name: import_zod.z.string().optional(),
          type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
          bytes: import_zod.z.number().optional(),
          format: import_zod.z.string().optional(),
          sha256: import_zod.z.string().optional(),
          width: import_zod.z.number().optional(),
          height: import_zod.z.number().optional(),
          order: import_zod.z.number().optional()
        }).optional().nullable(),
        attributes: import_zod.z.array(
          import_zod.z.object({
            trait_type: import_zod.z.string(),
            value: import_zod.z.union([import_zod.z.string(), import_zod.z.number()]),
            display_type: import_zod.z.string().optional()
          })
        ).optional().nullable(),
        media: import_zod.z.record(
          import_zod.z.object({
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]),
            url: import_zod.z.string(),
            name: import_zod.z.string().optional(),
            details: import_zod.z.object({
              name: import_zod.z.string().optional(),
              type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
              bytes: import_zod.z.number().optional(),
              format: import_zod.z.string().optional(),
              sha256: import_zod.z.string().optional(),
              width: import_zod.z.number().optional(),
              height: import_zod.z.number().optional(),
              order: import_zod.z.number().optional(),
              duration: import_zod.z.number().optional(),
              codecs: import_zod.z.array(import_zod.z.string()).optional(),
              loop: import_zod.z.boolean().optional()
            }).optional(),
            thumbnail: import_zod.z.object({
              url: import_zod.z.string(),
              details: import_zod.z.object({
                name: import_zod.z.string().optional(),
                type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
                bytes: import_zod.z.number().optional(),
                format: import_zod.z.string().optional(),
                sha256: import_zod.z.string().optional(),
                width: import_zod.z.number().optional(),
                height: import_zod.z.number().optional(),
                order: import_zod.z.number().optional()
              }).optional()
            }).optional(),
            poster: import_zod.z.object({
              url: import_zod.z.string(),
              details: import_zod.z.object({
                name: import_zod.z.string().optional(),
                type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
                bytes: import_zod.z.number().optional(),
                format: import_zod.z.string().optional(),
                sha256: import_zod.z.string().optional(),
                width: import_zod.z.number().optional(),
                height: import_zod.z.number().optional(),
                order: import_zod.z.number().optional()
              }).optional()
            }).optional()
          })
        ).optional().nullable(),
        royalties: import_zod.z.array(
          import_zod.z.object({
            address: import_zod.z.string(),
            percent: import_zod.z.number(),
            isPrimaryOnly: import_zod.z.boolean().optional()
          })
        ).optional().nullable(),
        customizing: import_zod.z.object({
          self: import_zod.z.object({
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]),
            url: import_zod.z.string(),
            name: import_zod.z.string().optional(),
            details: import_zod.z.object({
              name: import_zod.z.string().optional(),
              type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
              bytes: import_zod.z.number().optional(),
              format: import_zod.z.string().optional(),
              sha256: import_zod.z.string().optional(),
              width: import_zod.z.number().optional(),
              height: import_zod.z.number().optional(),
              order: import_zod.z.number().optional(),
              duration: import_zod.z.number().optional(),
              codecs: import_zod.z.array(import_zod.z.string()).optional(),
              loop: import_zod.z.boolean().optional()
            }).optional(),
            image_overlay_specs: import_zod.z.object({
              layer: import_zod.z.number().optional(),
              order_in_layer: import_zod.z.number().optional(),
              offset: import_zod.z.object({
                x: import_zod.z.number().optional(),
                y: import_zod.z.number().optional()
              }).optional(),
              opacity: import_zod.z.number().optional(),
              rotation: import_zod.z.number().optional(),
              scale: import_zod.z.object({
                x: import_zod.z.number().optional(),
                y: import_zod.z.number().optional(),
                unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
              }).optional(),
              anchor_point: import_zod.z.object({
                x: import_zod.z.number(),
                y: import_zod.z.number()
              }).optional(),
              parent_anchor_point: import_zod.z.object({
                x: import_zod.z.number(),
                y: import_zod.z.number()
              }).optional()
            }).optional(),
            placeholder: import_zod.z.object({
              url: import_zod.z.string(),
              details: import_zod.z.object({
                name: import_zod.z.string().optional(),
                type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
                bytes: import_zod.z.number().optional(),
                format: import_zod.z.string().optional(),
                sha256: import_zod.z.string().optional(),
                width: import_zod.z.number().optional(),
                height: import_zod.z.number().optional(),
                order: import_zod.z.number().optional()
              }).optional()
            }).optional(),
            tag: import_zod.z.string()
          }),
          slots: import_zod.z.record(
            import_zod.z.object({
              type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]),
              collections: import_zod.z.array(import_zod.z.union([import_zod.z.number(), import_zod.z.string()])).optional(),
              name: import_zod.z.string().optional(),
              image_overlay_specs: import_zod.z.object({
                layer: import_zod.z.number().optional(),
                order_in_layer: import_zod.z.number().optional(),
                offset: import_zod.z.object({
                  x: import_zod.z.number().optional(),
                  y: import_zod.z.number().optional()
                }).optional(),
                opacity: import_zod.z.number().optional(),
                rotation: import_zod.z.number().optional(),
                scale: import_zod.z.object({
                  x: import_zod.z.number().optional(),
                  y: import_zod.z.number().optional(),
                  unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
                }).optional(),
                anchor_point: import_zod.z.object({
                  x: import_zod.z.number(),
                  y: import_zod.z.number()
                }).optional(),
                parent_anchor_point: import_zod.z.object({
                  x: import_zod.z.number(),
                  y: import_zod.z.number()
                }).optional()
              }).optional()
            })
          ).optional(),
          mutator_reactions: import_zod.z.record(
            import_zod.z.object({
              layer: import_zod.z.number().optional().optional(),
              order_in_layer: import_zod.z.number().optional().optional(),
              offset: import_zod.z.object({
                x: import_zod.z.number().optional(),
                y: import_zod.z.number().optional()
              }).optional().optional(),
              opacity: import_zod.z.number().optional().optional(),
              rotation: import_zod.z.number().optional().optional(),
              scale: import_zod.z.object({
                x: import_zod.z.number().optional(),
                y: import_zod.z.number().optional(),
                unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
              }).optional().optional(),
              anchor_point: import_zod.z.object({
                x: import_zod.z.number(),
                y: import_zod.z.number()
              }).optional().optional(),
              parent_anchor_point: import_zod.z.object({
                x: import_zod.z.number(),
                y: import_zod.z.number()
              }).optional().optional(),
              url: import_zod.z.string().optional(),
              details: import_zod.z.object({
                name: import_zod.z.string().optional(),
                type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
                bytes: import_zod.z.number().optional(),
                format: import_zod.z.string().optional(),
                sha256: import_zod.z.string().optional(),
                width: import_zod.z.number().optional(),
                height: import_zod.z.number().optional(),
                order: import_zod.z.number().optional()
              }).optional()
            })
          ).optional(),
          mutators: import_zod.z.array(import_zod.z.string()).optional()
        }).optional().nullable(),
        customizingOverrides: import_zod.z.object({
          self: import_zod.z.object({
            type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
            url: import_zod.z.string().optional(),
            name: import_zod.z.string().optional().optional(),
            details: import_zod.z.object({
              name: import_zod.z.string().optional(),
              type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
              bytes: import_zod.z.number().optional(),
              format: import_zod.z.string().optional(),
              sha256: import_zod.z.string().optional(),
              width: import_zod.z.number().optional(),
              height: import_zod.z.number().optional(),
              order: import_zod.z.number().optional(),
              duration: import_zod.z.number().optional(),
              codecs: import_zod.z.array(import_zod.z.string()).optional(),
              loop: import_zod.z.boolean().optional()
            }).optional().optional(),
            image_overlay_specs: import_zod.z.object({
              layer: import_zod.z.number().optional(),
              order_in_layer: import_zod.z.number().optional(),
              offset: import_zod.z.object({
                x: import_zod.z.number().optional(),
                y: import_zod.z.number().optional()
              }).optional(),
              opacity: import_zod.z.number().optional(),
              rotation: import_zod.z.number().optional(),
              scale: import_zod.z.object({
                x: import_zod.z.number().optional(),
                y: import_zod.z.number().optional(),
                unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
              }).optional(),
              anchor_point: import_zod.z.object({
                x: import_zod.z.number(),
                y: import_zod.z.number()
              }).optional(),
              parent_anchor_point: import_zod.z.object({
                x: import_zod.z.number(),
                y: import_zod.z.number()
              }).optional()
            }).optional().optional(),
            placeholder: import_zod.z.object({
              url: import_zod.z.string(),
              details: import_zod.z.object({
                name: import_zod.z.string().optional(),
                type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
                bytes: import_zod.z.number().optional(),
                format: import_zod.z.string().optional(),
                sha256: import_zod.z.string().optional(),
                width: import_zod.z.number().optional(),
                height: import_zod.z.number().optional(),
                order: import_zod.z.number().optional()
              }).optional()
            }).optional().optional(),
            tag: import_zod.z.string().optional()
          }).optional(),
          slots: import_zod.z.record(
            import_zod.z.object({
              type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
              collections: import_zod.z.array(import_zod.z.union([import_zod.z.number(), import_zod.z.string()])).optional().optional(),
              name: import_zod.z.string().optional().optional(),
              image_overlay_specs: import_zod.z.object({
                layer: import_zod.z.number().optional(),
                order_in_layer: import_zod.z.number().optional(),
                offset: import_zod.z.object({
                  x: import_zod.z.number().optional(),
                  y: import_zod.z.number().optional()
                }).optional(),
                opacity: import_zod.z.number().optional(),
                rotation: import_zod.z.number().optional(),
                scale: import_zod.z.object({
                  x: import_zod.z.number().optional(),
                  y: import_zod.z.number().optional(),
                  unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
                }).optional(),
                anchor_point: import_zod.z.object({
                  x: import_zod.z.number(),
                  y: import_zod.z.number()
                }).optional(),
                parent_anchor_point: import_zod.z.object({
                  x: import_zod.z.number(),
                  y: import_zod.z.number()
                }).optional()
              }).optional().optional()
            })
          ).optional(),
          mutator_reactions: import_zod.z.record(
            import_zod.z.object({
              layer: import_zod.z.number().optional().optional().optional(),
              order_in_layer: import_zod.z.number().optional().optional().optional(),
              offset: import_zod.z.object({
                x: import_zod.z.number().optional(),
                y: import_zod.z.number().optional()
              }).optional().optional().optional(),
              opacity: import_zod.z.number().optional().optional().optional(),
              rotation: import_zod.z.number().optional().optional().optional(),
              scale: import_zod.z.object({
                x: import_zod.z.number().optional(),
                y: import_zod.z.number().optional(),
                unit: import_zod.z.enum(["px", "%"]).default("%").optional().optional()
              }).optional().optional().optional(),
              anchor_point: import_zod.z.object({
                x: import_zod.z.number(),
                y: import_zod.z.number()
              }).optional().optional().optional(),
              parent_anchor_point: import_zod.z.object({
                x: import_zod.z.number(),
                y: import_zod.z.number()
              }).optional().optional().optional(),
              url: import_zod.z.string().optional().optional(),
              details: import_zod.z.object({
                name: import_zod.z.string().optional(),
                type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
                bytes: import_zod.z.number().optional(),
                format: import_zod.z.string().optional(),
                sha256: import_zod.z.string().optional(),
                width: import_zod.z.number().optional(),
                height: import_zod.z.number().optional(),
                order: import_zod.z.number().optional()
              }).optional().optional()
            })
          ).optional(),
          mutators: import_zod.z.array(import_zod.z.string()).optional()
        }).optional().nullable(),
        animationUrl: import_zod.z.string().nullable(),
        animationDetails: import_zod.z.object({
          name: import_zod.z.string().optional(),
          type: import_zod.z.enum(["image", "animation", "video", "audio", "spatial", "pdf", "document", "other"]).optional(),
          bytes: import_zod.z.number().optional(),
          format: import_zod.z.string().optional(),
          sha256: import_zod.z.string().optional(),
          width: import_zod.z.number().optional(),
          height: import_zod.z.number().optional(),
          order: import_zod.z.number().optional()
        }).optional().nullable(),
        youtubeUrl: import_zod.z.string().nullable(),
        createdBy: import_zod.z.string().nullable(),
        backgroundColor: import_zod.z.string().nullable(),
        externalUrl: import_zod.z.string().nullable(),
        locale: import_zod.z.string().nullable(),
        tokenInfoError: import_zod.z.string().nullable(),
        isBurned: import_zod.z.boolean(),
        fetchedAtBlockNumber: import_zod.z.number(),
        updatedAtBlockNumber: import_zod.z.number(),
        updatedAt: import_zod.z.coerce.date(),
        createdAtBlockNumber: import_zod.z.number(),
        createdAt: import_zod.z.coerce.date()
      })
    ),
    count: import_zod.z.number()
  }),
  zSingleParsedNftParam: import_zod.z.object({
    collectionId: import_zod.z.string(),
    tokenId: import_zod.z.number()
  })
};
var tokenBalances = {
  zTokenBalancesFilter: import_zod.z.object({
    addressIn: import_zod.z.array(import_zod.z.string()).optional(),
    mode: import_zod.z.enum(["NFT", "Fungible", "RFT"]).optional(),
    collectionIdIn: import_zod.z.array(import_zod.z.string()).optional(),
    tokenIdIn: import_zod.z.array(import_zod.z.string()).optional(),
    amountFrom: import_zod.z.bigint().optional(),
    amountTo: import_zod.z.bigint().optional(),
    offset: import_zod.z.number().default(0).optional(),
    limit: import_zod.z.number().default(100).optional(),
    orderByAddress: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByCollectionId: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByTokenId: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByAmount: import_zod.z.enum(["asc", "desc"]).optional()
  }),
  zTokenBalancesWithCount: import_zod.z.object({
    items: import_zod.z.array(
      import_zod.z.object({
        key: import_zod.z.string(),
        address: import_zod.z.string(),
        collectionId: import_zod.z.number(),
        tokenId: import_zod.z.number(),
        amount: import_zod.z.string(),
        createdAtBlockNumber: import_zod.z.number(),
        updatedAtBlockNumber: import_zod.z.number(),
        fetchedAtBlockNumber: import_zod.z.number(),
        createdAt: import_zod.z.coerce.date(),
        updatedAt: import_zod.z.coerce.date()
      })
    ),
    count: import_zod.z.number()
  })
};
var tokensTransfers = {
  zTokenBalancesTransfersFilter: import_zod.z.object({
    fromIn: import_zod.z.array(import_zod.z.string()).optional(),
    toIn: import_zod.z.array(import_zod.z.string()).optional(),
    mode: import_zod.z.enum(["NFT", "Fungible", "RFT"]).optional(),
    collectionIdIn: import_zod.z.array(import_zod.z.string()).optional(),
    tokenIdIn: import_zod.z.array(import_zod.z.number()).optional(),
    amountFrom: import_zod.z.bigint().optional(),
    amountTo: import_zod.z.bigint().optional(),
    orderByFrom: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByTo: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByAmount: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByCollectionId: import_zod.z.enum(["asc", "desc"]).optional(),
    orderByTokenId: import_zod.z.enum(["asc", "desc"]).optional(),
    offset: import_zod.z.number().default(0).optional(),
    limit: import_zod.z.number().default(100).optional()
  }),
  zTokenBalancesTransfersWithCount: import_zod.z.object({
    items: import_zod.z.array(
      import_zod.z.object({
        id: import_zod.z.number(),
        from: import_zod.z.string(),
        to: import_zod.z.string(),
        collectionId: import_zod.z.number(),
        tokenId: import_zod.z.number(),
        amount: import_zod.z.string(),
        eventId: import_zod.z.number(),
        event: import_zod.z.object({
          blockNumber: import_zod.z.number(),
          blockTimestamp: import_zod.z.number(),
          extrinsicHash: import_zod.z.string(),
          extrinsic: import_zod.z.object({
            indexInBlock: import_zod.z.number(),
            section: import_zod.z.string().optional(),
            method: import_zod.z.string().optional(),
            signer: import_zod.z.string().optional()
          })
        })
      })
    ),
    count: import_zod.z.number()
  })
};

// src/indexer/routes.ts
var import_zod2 = require("zod");
var ROUTES = {
  accounts: {
    single: {
      route: "/accounts/:address",
      schemaIn: accounts.zSingleAccountParams,
      schemaOut: accounts.zAccountRowSelect
    },
    multiple: { route: "/accounts", schemaIn: accounts.zAccountsFilter, schemaOut: accounts.zAccountsWithCount }
  },
  blocks: {
    single: { route: "/blocks/:hashOrNumber", schemaIn: import_zod2.z.object({ hashOrNumber: import_zod2.z.union([import_zod2.z.string(), import_zod2.z.number().int()]) }), schemaOut: blocks.zBlockRowSelect },
    multiple: { route: "/blocks", schemaIn: blocks.zBlocksFilter, schemaOut: blocks.zBlocksWithCount }
  },
  coinBalances: {
    single: {
      route: "/balance/:address",
      schemaIn: coinBalances.zSingleCoinBalanceParam,
      schemaOut: coinBalances.zCoinBalanceRowSelect
    },
    multiple: { route: "/balance", schemaIn: coinBalances.zCoinBalancesFilter, schemaOut: coinBalances.zCoinBalancesWithCount }
  },
  balanceTransfers: {
    multiple: { route: "/balance/transfers", schemaIn: coinTransfers.zCoinBalancesTransfersFilter, schemaOut: coinTransfers.zCoinBalancesTransfersWithCount }
  },
  collections: {
    single: {
      route: "/collections/:collectionId",
      schemaIn: collections.zSingleCollectionParam,
      schemaOut: collections.zParsedCollectionRowSelect
    },
    multiple: { route: "/collections", schemaIn: collections.zCollectionsFilter, schemaOut: collections.zCollectionsWithCount }
  },
  events: {
    multiple: {
      route: "/events",
      schemaIn: events.zEventsFilter,
      schemaOut: events.zEventWithCount
    },
    sections: { route: "/events/sections", schemaIn: import_zod2.z.never(), schemaOut: events.zSectionMethodArray }
  },
  extrinsics: {
    single: {
      route: "/extrinsics/:hash",
      schemaIn: extrinsics.zSingleExtrinsicParam,
      schemaOut: extrinsics.zExtrinsicRowSelect
    },
    multiple: { route: "/extrinsics", schemaIn: extrinsics.zExtrinsicsFilter, schemaOut: extrinsics.zExtrinsicsWithCount },
    sections: { route: "/extrinsics/sections", schemaIn: import_zod2.z.never(), schemaOut: events.zSectionMethodArray }
  },
  parsedNfts: {
    single: {
      route: "/nfts/:collectionId/:tokenId",
      schemaIn: parsedNfts.zSingleParsedNftParam,
      schemaOut: parsedNfts.zParsedNftRowSelect
    },
    multiple: { route: "/nfts", schemaIn: parsedNfts.zParsedNftsFilter, schemaOut: parsedNfts.zParsedNftsWithCount }
  },
  tokenBalances: {
    multiple: {
      route: "/tokens",
      schemaIn: tokenBalances.zTokenBalancesFilter,
      schemaOut: tokenBalances.zTokenBalancesWithCount
    }
  },
  tokensTransfers: {
    multiple: {
      route: "/tokens/transfers",
      schemaIn: tokensTransfers.zTokenBalancesTransfersFilter,
      schemaOut: tokensTransfers.zTokenBalancesTransfersWithCount
    }
  }
};

// src/indexer/index.ts
var buildUrlWithParams = (url, params) => {
  let result = url;
  for (const key in params) {
    result = result.replace(`:${key}`, encodeURIComponent(params[key]));
  }
  return result;
};
var UniqueIndexer = (options) => {
  const client = import_axios.default.create({
    ...options.axiosConfig,
    baseURL: options.baseUrl,
    paramsSerializer: (params) => import_qs.default.stringify(params, { arrayFormat: "repeat" })
  });
  const requestViaPathParams = (routeConfig) => {
    const { route, schemaIn, schemaOut } = routeConfig;
    return async (input) => {
      const parseInputResult = schemaIn.safeParse(input);
      if (!parseInputResult.success)
        throw parseInputResult.error.format();
      const url = buildUrlWithParams(route, parseInputResult.data);
      const { data } = await client.get(url);
      return schemaOut.parse(data);
    };
  };
  const requestViaQueryString = (routeConfig) => {
    const { route, schemaOut } = routeConfig;
    const schemaIn = routeConfig.schemaIn.optional();
    return async (input) => {
      const params = schemaIn.parse(input);
      const { data } = await client.get(route, { params });
      return schemaOut.parse(data);
    };
  };
  return {
    options,
    block: requestViaPathParams(ROUTES.blocks.single),
    blocks: requestViaQueryString(ROUTES.blocks.multiple),
    account: requestViaPathParams(ROUTES.accounts.single),
    accounts: requestViaQueryString(ROUTES.accounts.multiple),
    coinBalance: requestViaPathParams(ROUTES.coinBalances.single),
    coinBalances: requestViaQueryString(ROUTES.coinBalances.multiple),
    balanceTransfers: requestViaQueryString(ROUTES.balanceTransfers.multiple),
    collection: requestViaPathParams(ROUTES.collections.single),
    collections: requestViaQueryString(ROUTES.collections.multiple),
    nft: requestViaPathParams(ROUTES.parsedNfts.single),
    nfts: requestViaQueryString(ROUTES.parsedNfts.multiple),
    events: requestViaQueryString(ROUTES.events.multiple),
    eventSections: requestViaQueryString(ROUTES.events.sections),
    extrinsic: requestViaPathParams(ROUTES.extrinsics.single),
    extrinsics: requestViaQueryString(ROUTES.extrinsics.multiple),
    extrinsicSections: requestViaQueryString(ROUTES.extrinsics.sections),
    tokenBalances: requestViaQueryString(ROUTES.tokenBalances.multiple),
    tokenTransfers: requestViaQueryString(ROUTES.tokensTransfers.multiple)
  };
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  UniqueIndexer
});
